WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift

```diff
@@ -0,0 +1,279 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import SnapshotTesting
+import XCTest
+
+@testable import WireAPI
+
+final class SelfUserAPITests: XCTestCase {
+
+    private var apiSnapshotHelper: APISnapshotHelper<SelfUserAPI>!
+
+    // MARK: - Setup
+
+    override func setUp() {
+        super.setUp()
+        apiSnapshotHelper = APISnapshotHelper { httpClient, apiVersion in
+            let builder = SelfUserAPIBuilder(httpClient: httpClient)
+            return builder.makeAPI(for: apiVersion)
+        }
+    }
+
+    override func tearDown() {
+        apiSnapshotHelper = nil
+        super.tearDown()
+    }
+
+    // MARK: - Request generation
+
+    func testGetSelfUserRequest() async throws {
+        try await apiSnapshotHelper.verifyRequestForAllAPIVersions { sut in
+            _ = try await sut.getSelfUser()
+        }
+    }
+
+    func testPushSupportedProtocolsRequest() async throws {
+        let supportedVersions = APIVersion.v5.andNextVersions
+        try await apiSnapshotHelper.verifyRequest(for: supportedVersions) { sut in
+            _ = try await sut.pushSupportedProtocols([.mls])
+        }
+    }
+
+    // MARK: - Request unsupported endpoint
+
+    func testPushSupportedProtocols_UnsupportedVersionError_V0() async throws {
```

Comment:
Personally I would do this for all versions where it is unsupported in a _single_ test. 

- As a form of documentation, it makes it easy to understand which API versions behave in the same way.
- It lowers the effort to test all API versions thoroughly.
- It saves a lot of boilerplate.

You can have an array of `APIVersion`s and build the `sut` using `SelfUserAPIBuilder`. While this kind of parameterized testing is not that common in iOS at the moment it will become common when we move to swift testing.

Reply:
Good point 👍  this way when we introduce a new API version, we make sure new versioned APIs are properly tested on the existing endpoints without having to write additional tests (as you mentioned), I did some changes just for SelfUser for now but it could also be nice to generalize and automatize this behavior to all APIs, we should keep this in mind.